회원가입-저장구현 및 회원아이디 중복체크

Jeonminje opened the pull request:
save()를 이용한 회원정보 저장.
findByLoginId()를 이용한 회원아이디 중복 여부 확인.

Jeonminje committed:
회원저장 및 회원아이디 중복확인

save()를 이용한 회원정보 저장.
findByLoginId()를 이용한 회원아이디 중복 여부.

Jeonminje committed:
SignUp-emailCheck

사용자 이메일로 메일 전송 및 전송코드 반환

Jeonminje committed:
SignUp-login

로그인 및 로그아웃을 세션을 이용해 구현

sunwootest commented:
타겟브랜치 변경해주세요

sunwootest commented on movieservice/src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+import study.movieservice.service.MemberService;
+
+
+
+@Controller
```
json 데이터를 전송할텐데 @Controller로 충분할까요?

sunwootest commented on movieservice/src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+
+@Controller
+@RequiredArgsConstructor
+@RequestMapping("/member")
```
보통 복수형을 사용합니다. API 명명규칙에 대해서 자세히 알아보세요 :)

sunwootest commented on movieservice/src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+    @ResponseBody
+    @PostMapping("/idcheck") // 아이디 중복 확인
+    public String loginIdCheck(@RequestBody Member member) {
+        if (memberService.checkLoginId(member.getLoginId())) {
+            return "이미 사용중인 아이디 입니다.";
```
이미 사용중인 아이디여도 200 http 상태코드를 보내주겠네요 :)

sunwootest commented on movieservice/src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+@Controller
+@RequiredArgsConstructor
+@RequestMapping("/member")
+@Slf4j
```
해당 어노테이션은 왜 작성하셨나요?

Jeonminje replied:
스프링 빈이 잘 주입되는지 확인하기 위해 사용하였는데 지금은 사용하지않아 지우도록 하겠습니다!

sunwootest commented on movieservice/src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+
+    @ResponseBody
+    @PostMapping("/idcheck") // 아이디 중복 확인
+    public String loginIdCheck(@RequestBody Member member) {
```
id의 중복여부만 확인하는데 Member를 RequestBody로 받아올 필요가 있을까요?

Jeonminje replied:
수정하겠습니다!

sunwootest commented on movieservice/src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+    @PostMapping("/join") // 회원가입
+    public String memberJoin(@RequestBody Member member) {
+        memberService.addMember(member);
+        return "ok";
```
성공하면 200 상태코드만 보내줘도 충분합니다.

sunwootest replied:
ExeptionHandler로 Http 상태코드를 전송하는 방법을 사용해도 좋을 것 같습니다.

sunwootest replied:
ok를 꼭 return 해야 할까요?

sunwootest commented on movieservice/src/main/java/study/movieservice/domain/member/Member.java:
```diff
@@ -0,0 +1,22 @@
+    private Grade grade;
+
+
+    public void setGrade(Grade grade) {
+        this.grade = grade;
+    }
+
+    public void setLoginPassword(String loginPassword) {
+        this.loginPassword = loginPassword;
+    }
```
불변객체가 아니네요 이렇게 되면?

sunwootest commented on movieservice/src/main/java/study/movieservice/repository/MemberRepository.java:
```diff
@@ -0,0 +1,10 @@
+package study.movieservice.repository;
+
+import study.movieservice.domain.member.Member;
+
+import java.util.Optional;
+
+public interface MemberRepository {
+    void save(Member member);
+    Optional<Member> findByLoginId(String loginId);
+}
```
MyBatis는 Mapper만 있으면 충분합니다.

sunwootest commented on movieservice/src/main/java/study/movieservice/repository/mapper/MemberMapper.java:
```diff
@@ -0,0 +1,12 @@
+@Mapper
+public interface MemberMapper {
+    public void save(Member member);
+    public Optional<Member> findByLoginId(String loginId);
```
Optional은 왜 사용하셨나요?

Jeonminje replied:
로그인 아이디 중복여부를 확인하기 위해 사용하였는데 테이블에 저장되어있지 않은 경우를 고려해 Optional을 이용해 null처리를 하였습니다

sunwootest commented on movieservice/src/main/java/study/movieservice/service/MemberService.java:
```diff
@@ -0,0 +1,43 @@
+    public void addMember(Member member){
+        String encodingPassword = BCrypt.hashpw(member.getLoginPassword(), BCrypt.gensalt());
+
+        member.setLoginPassword(encodingPassword);
+        member.setGrade(Grade.BASIC);
```
불변객체가 아니게 되었네요

Jeonminje replied:
수정하겠습니다!

sunwootest commented on movieservice/src/main/resources/application.properties:
```diff
@@ -0,0 +1,12 @@
+#?? ??? ??.
+
+spring.datasource.url =jdbc:mariadb://localhost:3303/movie_service
+spring.datasource.username=root
+spring.datasource.password=alsdn120
+
+#MyBatis in xml convenient resultType. ( domian ??? ?? ?? )
+mybatis.type-aliases-package=study.movieservice.domain
+#db and java convert snake-case, camel-case
+mybatis.configuration.map-underscore-to-camel-case=true
+#session life
+server.servlet.session.timeout=60
```
이 정보를 모두 공개해도 될까요?

Jeonminje replied:
삭제한뒤 .gitignore에 올려 앞으로 커밋에서 제외시키도록하겠습니다.

sunwootest commented on movieservice/src/main/java/study/movieservice/service/MemberService.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public boolean checkLoginId(String loginId){
+        Optional<Member> finding  = memberRepository.findByLoginId(loginId);
+
```
id 존재유무를 확인하는데 Member 정보를 전체 불러올 필요가 있을까요? SQL 문법 중에 EXISTS가 있는데 참고해보세요.

Jeonminje replied:
넵!

sunwootest commented on movieservice/src/main/resources/study/movieservice/repository/mapper/MemberMapper.xml:
```diff
@@ -0,0 +1,16 @@
+        "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
+<mapper namespace="study.movieservice.repository.mapper.MemberMapper">
+    <insert id="save" useGeneratedKeys="true" keyProperty="memberId">
+        insert into member(email, login_id, login_password, nickname, grade)
+        values (#{email}, #{loginId}, #{loginPassword},#{nickname},#{grade})
```
created_at, updated_at에 관련된 정보가 안보이네요.

Jeonminje replied:
테이블에 default값으로 설정해 놓았습니다!

sunwootest reviewed (commented):
타겟브랜치 변경해주시고 브랜치 명명 규칙도 통일해주세요. 다른 프로젝트에서 브랜치 명은 어떻게 정하는지도 알아보시면 좋을 것 같습니다.

Jeonminje committed:
application.properties 삭제

앞으로 보안을 위해 커밋을 하지않을 예정

Jeonminje committed:
회원정보 저장 및 아이디 중복확인 기능

DTO를 이용해 회원정보 저장 및 아이디 중복확인 코드를 수정했습니다.

Jeonminje commented:
커밋했습니다!

sunwootest commented on movieservice/src/main/java/study/movieservice/domain/member/Member.java:
```diff
@@ -0,0 +1,16 @@
+    private final String loginPassword;
+    private final String nickname;
+    private final Grade grade;
+
+}
```
컨벤션 관련 실수를 해결하려면 어떻게 해야할까요?

Jeonminje replied:
관용적인 방법에 따라 좀 더 신경쓰도록 하겠습니다!

sunwootest replied:
인간은 누구나 실수하기 마련입니다. 좀 더 확실하게 실수를 줄일 수 있는 안전장치를 알아보시면 좋을 것 같습니다.

sunwootest commented on movieservice/src/main/java/study/movieservice/service/MemberService.java:
```diff
@@ -0,0 +1,47 @@
+    public boolean checkLoginId(String loginId){
+        Optional<LoginIdDTO> finding  = memberRepository.findByLoginId(loginId);
+
+        if(finding.isPresent()){ // 값이 있다면.
```
해당 주석이 꼭 필요할까요?

Jeonminje replied:
불필요한 주석이므로 제거하도록 하고
주석이 필요할 시 javadoc의 대상이 되는 문서화 주석을 이용하겠습니다.

sunwootest commented on movieservice/src/main/java/study/movieservice/service/MemberService.java:
```diff
@@ -0,0 +1,47 @@
+
+    public void addMember(MemberDTO memberDTO){
+        String encodingPassword = BCrypt.hashpw(memberDTO.getLoginPassword(), BCrypt.gensalt());
+        Member member = new Member(
+                memberDTO.getEmail(),
+                memberDTO.getLoginId(),
+                encodingPassword,
+                memberDTO.getNickname(),
+                Grade.BASIC);
```
Builder 패턴으로 개선할 수 있어 보이네요 :)

Jeonminje replied:
수정하겠습니다!

Jeonminje committed:
- /members/id-check
 loginId를 받기위해 기존 @RequestBody와 loginDTO에서
@RequestParam을 이용해 String으로 받아 주었습니다.

loginId가 이미 테이블에 있을 시,
exceptionHandler를 이용한 처리로 변경했습니다.

- MemberRepository
MemberMapper와 역할이 중복된다 여겨 MemberRepository 클래스 및 인터페이스를 삭제했습니다.

- /members/join
/members/sign-up으로 url변경했습니다.

- MemberMapper.xml
기존 Optional을 이용해 Member 객체를 받던것에서
SQL을 수정해 select exists문으로 boolean 타입을 리턴해 좀더 성능을 높였습니다.

- MemberService.addMember()
롬복을 이용해 builder 패턴을 적용하여 좀 더 가시적으로 코드를 수정했습니다.

- 파일 수정
기존 깃에 대한 이해 부족으로인한
파일 중복 문제 등을 해결하였습니다.

Jeonminje commented:
커밋 했습니다!
약간의 코드변경과 MemberRepository 클래스 및 인터페이스 삭제,
기존 application.properties파일 및 다른 파일 등의 중복문제를 해결 하다보니 파일변경이 좀 많은 것 같습니다.. 죄송합니다

sunwootest commented on src/main/java/study/movieservice/service/MemberService.java:
```diff
@@ -0,0 +1,42 @@
+
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
```
사용하지 않는 import문도 제거해주세요. 인텔리제이에서 설정할 수 있습니다 :)

Jeonminje replied:
네 알겠습니다!

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -0,0 +1,38 @@
+public class MemberController {
+    private final MemberService memberService;
+
+    @PostMapping("/id-check")
+    public ResponseEntity<String> loginIdCheck(@RequestParam("login_id") String loginId) throws LoginException {
```
RequestParam으로 데이터를 받아올 때 snakecase를 사용하지는 않습니다.

Jeonminje replied:
넵 수정하겠습니다

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -0,0 +1,38 @@
+        memberService.addMember(memberDTO);
+    }
+
+    @ExceptionHandler
+    public ResponseEntity<String> loginHandle(LoginException e) {
+        return new ResponseEntity<>(e.getMessage(), HttpStatus.CONFLICT);
+    }
```
Controller에 ExceptionHandler 메소드도 들어가 있네요. 클래스를 분리하는 편이 나아보입니다.

Jeonminje replied:
수정하겠습니다!

sunwootest commented on src/main/java/study/movieservice/service/MemberService.java:
```diff
@@ -0,0 +1,42 @@
+
+    private final MemberMapper memberMapper;
+
+    public void addMember(MemberDTO memberDTO){
+        String encodingPassword = BCrypt.hashpw(memberDTO.getLoginPassword(), BCrypt.gensalt());
+        Member member = Member.builder()
+                .email(memberDTO.getEmail())
+                .loginId(memberDTO.getLoginId())
+                .loginPassword(encodingPassword)
+                .nickname(memberDTO.getNickname())
+                .grade(Grade.BASIC).build();
+        memberMapper.save(member);
+    }
```
아이디와 비밀번호의 규정에 대한 제한장치가 없어보이네요. 해당 내용은 다른 이슈로 분리해서 처리하면 좋을 것 같습니다. 스프링에서는 어떻게 해결할 수 있는지 알아보세요 :)

Jeonminje replied:
넵 추후에 다른 이슈로 분리해서 처리하겠습니다!

sunwootest commented on src/main/resources/study/movieservice/repository/MemberMapper.xml:
```diff
@@ -0,0 +1,18 @@
+        select exists (
+            select login_id
+            from members
+            where login_id = #{loginId}
+                   )
```
사소하지만 SQL 문도 불필요한 공백 주의해주세요!

Jeonminje replied:
넵!

sunwootest commented on src/main/java/study/movieservice/domain/member/MemberDTO.java:
```diff
@@ -0,0 +1,16 @@
+
+@Getter
+@RequiredArgsConstructor
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
```
JsonNaming 어노테이션은 왜 작성하셨나요?

Jeonminje replied:
해당 클래스는 json 데이터를 받는 DTO 클래스이기 때문에 혹시모를 상황에 자바표준인 camelCase로 변환하기위해 작성하였습니다.

sunwootest replied:
Json이 정말 snakeCase로 작성되는지 다시 한 번 확인해보세요 :)

Jeonminje replied:
넵 감사합니다!

sunwootest reviewed (commented):
조금만 더 다듬으면 Approve 받을 수 있을 것 같네요 :)

Jeonminje committed:
GeneralExceptionHandler 클래스 추가

기존 컨트롤러안에서 예외를 컨트롤 하던 것에서
따로 예외를 처리해주는 클래스를 생성해 그안에서 모두 처리하도록 해주었습니다.

인텔리제이의 사용하지않는 import문을 제거하도록 기능을 설정했습니다.

@RequestParam으로 데이터를 받을 떄 snakeCase로 받던 부분을 수정했습니다.

Jeonminje commented:
커밋했습니다!

Jeonminje committed:
불필요한 import

불필요한 import 제거

sunwootest reviewed (commented):
JSON 관련 코드리뷰만 수정해주세요. 몇가지 더 다듬을 부분이 존재하긴 하지만 일단 사실상 approve 하겠습니다 :) 

다른 팀원과 짝 프로그래밍 진행중이기 때문에 아직 merge는 하지 말아주세요.

Jeonminje commented:
넵 감사합니다

Jeonminje committed:
application.properties 파일 삭제

보안을 위해 삭제하였습니다.

Jeonminje committed:
Merge branch 'SignUp-save-minje' into SignUp-emailCheck-minje

# Conflicts:
#	build/classes/java/main/study/movieservice/config/MailConfig.class
#	build/classes/java/main/study/movieservice/service/MailSender.class
#	build/tmp/compileJava/compileTransaction/stash-dir/MemberController.class.uniqueId0
#	movieservice/build/classes/java/main/study/movieservice/controller/MemberController.class
#	movieservice/build/tmp/compileJava/previous-compilation-data.bin
#	movieservice/src/main/java/study/movieservice/controller/MemberController.java
#	src/main/java/study/movieservice/config/MailConfig.java
#	src/main/java/study/movieservice/service/MailSender.java

Jeonminje committed:
Merge remote-tracking branch 'origin/SignUp-emailCheck-minje' into SignUp-emailCheck-minje

Jeonminje committed:
이메일 인증코드 전송 수정 및 코드 확인 로직 추가

- 기존 JavaMailSender를 설정한 뒤 Bean으로 수동 등록 하던 것에서
properties 파일을 이용해 설정을 숨기고,
JavaMailSender는 스프링이 등록 해주는 것을 사용하는 방향으로 수정하였습니다.

- controller.GeneralExceptionHandler 클래스
이메일 전송 에러 처리 추가했습니다.

- controller.MemberController 클래스
이메일 전송 및 이메일 코드 확인 관련 매핑 추가했습니다.

- repository.RedisUtil 클래스 추가
이메일 코드를 보관하기 위해 redis를 사용하였으며, key는 이메일, value는 코드를 사용하고
유효시간을 설정하도록 하였습니다.
getData() 함수는 Optional을 반환하여 데이터가 없을 시,
호출한 MailService에서 null처리를 해주도록하였습니다.

- service.MailService 클래스
createMessage() 에서 메시지내용과 제목을 직접 작성하던 것을 따로 함수로 작성하였습니다.
필드에 있던 인증코드 변수를 동시성 문제를 고려해 함수안으로 넣었고,
함수간에 파라미터로 생성한 인증코드를 전달 해 사용하는 방식으로 변환했습니다.
checkEmailCode() 함수를 추가해 전송한 코드와 일치하는지 확인하도록 하였습니다.
추가로 111111~999999였던 코드 범위를 000000~999999 범위까지 늘렸습니다.

Jeonminje committed:
잘못한 커밋 삭제

예전에 잘못 커밋한 파일입니다.

Jeonminje committed:
잘못한 커밋 삭제

잘못된 커밋입니다.

Jeonminje committed:
잘못된 커밋 삭제

보안관련 문제로 삭제하겠습니다.

Jeonminje committed:
-MemberController
 MailService에서 예외처리로 변경

- RedisUtil
 삭제 후 MailService내에서 StringRedisTemplate 주입받아 사용

- MailService
 기존 발신자 정보를 프로퍼티파일로 분리했습니다.
 createMessage()에서 던지던 예외를 unchecked예외로 변환 후 던져 ExceptionHandler에서 처리했습니다.
 Transactional 어노테이션 삭제했습니다. (redis는 rollback을 지원하지 않고, 인증코드는 3분뒤에 사라지고 또한 사용자는 메시지 전송 실패를 전달 받아 인증코드 재발급 받을 시 overwrite되므로 상관없다 생각했습니다.)
 기존 MailService에서 작성하던 메시지 내용 및 제목을 MessageForm 클래스를 만들어 더이상 의존하지 않도록 처리했습니다.

Jeonminje committed:
# 관심사 분리

## MemberController
- memberEmailCheck() 함수에서 mailService 클래스의 sendMail() 를 사용하던것에서 MemberService 클래스에 sendSignUpMessage()함수를 사용하여 회원가입 인증코드 전송.

## MailService
- address 와 sender 변수를 private final로 바꾸고 생성자의 파라미터에서 @value를 통해 properties파일에 있는 데이터를 주입해주었습니다.
- 기존 createMessage() 함수를 createSignUpMessage() 함수로 변경해 회원가입시 사용하는 메시지를 생성해주는 함수로 지정해주고, 메시지 생성에 문제가 없다면 storeKey() 를 통해 레디스에 이메일과 코드를 저장해주었습니다.
- 기존 sendMail() 함수가 회원가입 관심사에 의존적이다 판단해, MimeMessage 인스턴스를 생성해 넘겨주면 단순히 전송해주는 역할로 바꾸었습니다.

## MemberService
- 필드에 MailService를 주입받아 사용하는 것이 추가되었습니다.
- sendSignUpMessage() 함수를 추가하였습니다. MailService 클래스의 함수를 이용해 인증코드 생성, MimeMessage 객체생성 및 전송을 담당하는 역할 입니다.

## MessageForm
- 함수명을 messageSubject() -> signUpMessageSubject() , messageContent() -> signUpMessageContent() 로 책임에 따라 함수명을 지어줬습니다.

Jeonminje committed:
Merge branch 'SignUp-emailCheck-minje' into SignUp-login-minje

# Conflicts:
#	movieservice/src/main/java/study/movieservice/service/MemberService.java
#	src/main/java/study/movieservice/controller/LoginController.java
#	src/main/java/study/movieservice/domain/SessionConst.java

Jeonminje committed:
로그인 기능 수정 및 이메일 중복확인 기능과 merge

## LoginController
- 로그인과 로그아웃 요청이 들어오면 MemberService의 함수로 처리하였습니다.

## MemberMapper
- getMember() 함수를 추가하였습니다.

## MemberService
- logIn() logOut() 함수를 추가해 세션을 이용한 로그인, 로그아웃을 구현하였습니다.

Commit:
지켜지지 않은 코드 컨벤션 수정

Commit:
Merge branch 'SignUp-emailCheck-minje' into SignUp-login-minje

Commit:
로그인 컨트롤러 및 로그인 서비스 관련 함수 수정

## LoginController
- HttpServletRequest에 의존하던 코드를 HttpSession을 직접 받아 사용하는 것으로 변경하였습니다.
- 로그인 실패시 200코드를 반환해주던 것에서 오류코드를 리턴해주도록 변경하였습니다.
- 아이디와 비밀번호를 받기위해 LoginParam 클래스를 추가해 DTO로 이용하였습니다.

## LoginParam
- 아이디와 비밀번호를 받기위한 DTO로 이번에 추가했습니다.

## MemberMapper.xml
- 테이블 명을 members -> member로 바꾸고 관련 쿼리를 수정하였습니다.

## SessionConst
- 세션의 키값을 가지고 있는 클래스를 만들어 주었습니다. (닉네임, 로그인아이디)

## MemberService
- logIn()에서 기존 세션에 key = 아이디, value = 닉네임 으로 저장하던 것에서 각각 value로 저장하도록 수정했습니다. 또한 기존 HttpServletRequest 관련 코드를 HttpSession으로 수정하였습니다.
- logOut() : 기존 loginId를 파라미터로 받아 로그인아이디가 key로 저장된 세션에서 삭제를 해주는 방식에서 HttpSession.invalidate() 함수를 이용해 로그아웃 시 세션을 비우는 방식으로 수정하였습니다.

Jeonminje committed:
예외 추가

## exception 폴더를 만들어 예외 추가 ( 모두 런타임 예외 )
- LoginIncorrectException : 아이디나 비밀번호가 일치하지 않을 경우 발생하는 예외 입니다.
- LoginSetAttributeException : 세션의 key, value를 세팅할 때 예외가 발생하면 이 예외로 변환을 해줍니다.
- LogoutException : 로그아웃시 발생하는 예외 입니다.

## LoginController
- 로그인이 실패하면 예외를 발생시켜 ExceptionHandler에서 응답을 하도록 해주었습니다.

## MemberMapper
- getMember() 함수의 명을 수정하였습니다.

## MemberService
- logOut(), logIn() 함수에서 HttpSession 관련 로직을 분리하였습니다. SessionManager를 빈으로 주입받아 사용 합니다.

## SessionManager
- HttpSession을 빈으로 주입받아 세션관련 로직을 수행해줍니다.

Jeonminje committed:
GeneralExceptionHandler 추가 및 예외 처리 수정
1. 기존 controller에서 ResponseEntity 를 리턴하던 것에서 String을 리턴하는 것으로 방식을 바꾸고 @ResponseStatus 어노테이션을 이용하여 상태를 응답해주었습니다.
2. controller에 있던 비즈니스 로직을 모두 제거하였습니다. ( 예외는 모두 service단에서 던져주었습니다.)
3. EexceptionMessageConst 이넘 클래스를 생성하여 예외 메시지를 분리해주었습니다.
4. 기존 RuntimeException을 상속하여 직접 만들어 주었던 예외들을 모두 삭제, 및 만들어져 있는 예외를 사용하는 방안으로 변환해주었습니다.
5. 로그인 시, 세션에 저장하는 값을 로그인아이디가 아닌 primary key 값인 memberId를 저장해주었습니다.

Jeonminje committed:
log-in시 session에 nickname과 memberId를 저장하던 것을 닉네임은 저장하지 않는 코드로 수정.

Jeonminje committed:
SessionConst의 nickname관련 상수 제거

Jeonminje committed:
Merge pull request #11 from Jeonminje/SignUp-login-minje

로그인 기능

Jeonminje committed:
Merge pull request #8 from Jeonminje/SignUp-emailCheck-minje

회원가입-이메일인증
